lib/spack/spack/package.py

```diff
@@ -1428,6 +1578,114 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class Package(PackageBase):
+    phases = ['install']
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class EditableMakefile(PackageBase):
+    phases = ['edit', 'build', 'install']
+
+    def wdir(self):
+        return self.stage.source_path
+
+    def build_args(self):
+        return list()
+
+    def install_args(self):
+        return list()
+
+    def edit(self, spec, prefix):
+        raise NotImplementedError('\'edit\' function not implemented')
+
+    def build(self, spec, prefix):
+        args = self.build_args()
+        with working_dir(self.wdir()):
+            inspect.getmodule(self).make(*args)
+
+    def install(self, spec, prefix):
+        args = self.install_args() + ['install']
+        with working_dir(self.wdir()):
+            inspect.getmodule(self).make(*args)
+
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
+        return list()
+
+    def configure(self, spec, prefix):
+        options = ['--prefix={0}'.format(prefix)] + self.configure_args()
+        inspect.getmodule(self).configure(*options)
+
+    def build(self, spec, prefix):
+        inspect.getmodule(self).make()
+
+    def install(self, spec, prefix):
+        inspect.getmodule(self).make('install')
+
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class CMakePackage(PackageBase):
+    phases = ['cmake', 'build', 'install']
+
+    def build_type(self):
+        return 'RelWithDebInfo'
+
+    @property
+    def std_cmake_args(self):
+        # standard CMake arguments
+        args = ['-DCMAKE_INSTALL_PREFIX:PATH={0}'.format(self.prefix),
```

Comment:
They were in lib/spack/spack/build_environment.py for the record. I don't think we can remove these from build_environment without breaking backwards compatibility. But it would be nice if they weren't duplicated. Otherwise, when you update one, it might not be the right one.